python/addons/preprocessors.py

```diff
@@ -0,0 +1,119 @@
+        super(TensorFlowPreprocessor, self).__init__(feature, vectorizer, index, vocab, **kwargs)
+        self.FIELD_NAME = kwargs.get('FIELD_NAME', 'tokens')
+        self.upchars = tf.constant([chr(i) for i in range(65, 91)])
+        self.lchars = tf.constant([chr(i) for i in range(97, 123)])
+        self.upchars_lut = tf.contrib.lookup.index_table_from_tensor(mapping=self.upchars, num_oov_buckets=1, default_value=-1)
+
+    def lowercase(self, raw_post):
+        split_chars = tf.string_split(tf.reshape(raw_post, [-1]), delimiter="").values
+        upchar_inds = self.upchars_lut.lookup(split_chars)
+        return tf.reduce_join(tf.map_fn(lambda x: tf.cond(x[0] > 25,
+                                                          lambda: x[1],
+                                                          lambda: self.lchars[x[0]]),
+                                        (upchar_inds, split_chars), dtype=tf.string))
+
+    def preproc(self, tf_example):
+        """
+        Create a preprocessor chain inside of the tensorflow graph.
+        """
+        pass
+
+    def resize_sen(self, raw):
+        """
+        Splits and rejoins a string to ensure that tokens meet
+        the required max len.
+        """
+        raw_tokens = tf.string_split(tf.reshape(raw, [-1])).values
+        # sentence length > mxlen
+        raw_post = tf.reduce_join(raw_tokens[:self.mxlen], separator=" ")
+        return raw_post
+
+    @staticmethod
+    def reshape_indices(indices, shape):
+        reshaped = tf.sparse_reset_shape(indices, new_shape=shape)
+        # Now convert to a dense representation
+        x = tf.sparse_tensor_to_dense(reshaped)
+        x = tf.contrib.framework.with_shape(shape, x)
+        return x
+
+
+@exporter
+@register_preprocessor(name='token1d')
+class Token1DPreprocessor(TensorFlowPreprocessor):
+
+    def __init__(self, feature, vectorizer, index, vocab, **kwargs):
+        super(Token1DPreprocessor, self).__init__(feature, vectorizer, index, vocab, **kwargs)
+        self.mxlen = self.vectorizer.mxlen
+
+    def create_word_vectors_from_post(self, raw_post, lowercase=True):
+        # vocab has only lowercase words
+        word2index = self.index
+        if lowercase:
+            raw_post = self.lowercase(raw_post)
+        word_tokens = tf.string_split(tf.reshape(raw_post, [-1]))
+        word_indices = word2index.lookup(word_tokens)
+        # Reshape them out to the proper length
+        reshaped_words = tf.sparse_reshape(word_indices, shape=[-1])
+        return self.reshape_indices(reshaped_words, [self.mxlen])
+
+    def preproc(self, post_mappings):
+        # Split the input string, assuming that whitespace is splitter
+        # The client should perform any required tokenization for us and join on ' '
+        raw_post = post_mappings[self.FIELD_NAME]
+        raw_post = self.resize_sen(raw_post)
+        return self.create_word_vectors_from_post(raw_post)
```

Comment:
Token1d should check the `self.vectorizer.transform_fn.__name__` do decide if it should lowercase or not.

If this function is not lowercase or identity it should probably raise an error that it can't export arbitrary functions.

Reply:
agree with the first part, the second i would think is the designer's responsibility.

Reply:
By designer you mean who ever is writing a plugin? I think that makes sense but in this case the designer is you and I think for our use cases a safe guard like this makes sense, especially because it's not super easy to tell after exporting if it is lowercasing or not